src/routes/adm/csvRegisterUser.js

```diff
@@ -0,0 +1,138 @@
+ * @memberof module:router~mainRouter~userRouter~csvRegisterUser
+ * @inner
+ * @async
+ * @param {Object} req
+ * @param {Object} res
+ * @returns 200 if ok
+ * @returns 422 if invalid csv
+ * @returns 500 if Internal Server Error
+ */
+module.exports = async (req, res) => {
+  try {
+    // Verify received data
+    if (req.file.size === 0) { return res.status(422).json({ message: 'The file is empty' }) }
+
+    const csv = await convertCsv(req.file.path)
+    if (!csv || csv.length === 0) { return res.status(422).json({ message: 'Invalid request' }) }
+
+    if (checkCsvHeader(csv[0])) { return res.status(422).json({ message: 'Csv header is not valid' }) }
+
+    const error = await checkCsvBody(csv)
+    if (error.length !== 0) {
+      return res.status(422).json(error)
+    }
+
+    const [err, line, row] = await processImport(csv)
+    if (err) {
+      return res.status(422).json({
+        line,
+        row
+      })
+    }
+
+    return res.status(200).json('ok')
+  } catch (error) /* istanbul ignore next */ {
+    console.error(error)
+    return res.status(500).json({ message: 'Internal Server Error' })
+  }
+}
+
+const convertCsv = async (filepath) => {
+  const csv = []
+
+  const promise = new Promise((resolve, _) => {
+    try {
+      fs.createReadStream(filepath)
+        .pipe(csvParser())
+        .on('data', (data) => csv.push(data))
+        .on('end', () => {
+          resolve(csv)
+        })
+    } catch (e) {
+      resolve(null)
+    }
+  })
+
+  return await promise
+}
+
+const checkCsvHeader = (row) => {
+  const keys = ['firstname', 'lastname', 'email', 'role', 'class']
+
+  for (const key of Object.keys(row)) {
+    if (!keys.includes(key)) { return true }
+  }
+  return false
+}
+
+const checkCsvBody = async (csv) => {
+  const error = []
+
+  const addError = (errorType, index) => {
+    const idx = error.findIndex((e) => e.rowCSV === index + 2)
+    if (idx === -1) { error.push({ rowCSV: index + 2, errors: [errorType], ...csv[index] }) } else { error[idx].errors.push(errorType) }
+  }
+
+  for (const [index, row] of csv.entries()) {
+    if (row.firstname.length === 0 || !/^([a-zA-Z]| |-)+$/.test(row.firstname)) addError('Firstname is not valid', index)
+    if (row.lastname.length === 0 || !/^([a-zA-Z]| |-)+$/.test(row.lastname)) addError('Lastname is not valid', index)
+    if (row.email.length === 0 || !/^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/.test(row.email)) addError('Email is not valid', index)
+    if (row.role.length === 0 || !['student', 'teacher', 'adm'].includes(row.role.toLowerCase())) addError('Role is not valid', index)
+    if (row.class.length === 0 || !/^([a-zA-Z0-9]| |-)+$/.test(row.class)) addError('Class is not valid', index)
+
+    if (await Users.findOne({ email: row.email })) {
+      addError('User already exists', index)
+    }
+
+    const class_ = Classes.findOne({ name: row.class })
+
+    if (!class_ || class_.length === 0) {
+      addError('Invalid class', index)
+    }
+  }
```

Comment:
Student can only have one class.

Teacher can have one or more

You need to check that